### ISSUE TITLE
POST body is not inspected by modsecurity

### ISSUE THREAD

**Issue by code1955** [SpiderLabs/ModSecurity#684]
Hi,

We are suffering an issue related to POST inspection.
We are running modsecurity 2.7.7 on apache 2.5.3
Backend application is running on Jetty 2.9.
The application presents a login form to the end user.
When filling in the login fields with crafted data, like sql strings, the call is accepted and sent to the backend application.
But adding sql strings in the URL blocks the call.
Backend application is at risk as modsecurity is the only security control in place.
I've been suggested to enable SecStreamInBodyInspection, but it doesn't work.
Any help is welcome.

Kind regards,

**Comment by zimmerle** [comment]
Hi @code1955, yesterday I've merge #904 into nginx_refactoring branch. It should fix this issue, please confirm that the issue is fixed.

**Comment by lelekos** [comment]
Hello Guys,

I'm trying to come up with a simple rule that will deny specific string on a POST body data.

Can you help me create this rule? :)

Thanks
A.

**Comment by void-in** [comment]
@lelekos You shouldn't comment on existing threads your issue is not related. Also I would suggest you read the ModSecurity reference manual or use the mailing list for support queries.

Anyway, a simple rule like:
```
SecRule REQUEST_BODY "bad-string-to-block" "id:00001,deny,msg:'bad request blocked'"
```

**Comment by code1955** [comment]
Hi Zimmerle , and others, 

Reading back this post, I apologise for my silence 
I'm not any more working for the company where issue occurred, since 2013.
So I'm unable providing any feedback,  sorry. 
Nevertheless I'd like to thank you for your commitments, and I hope this post was useful for others .
Good job guys, don't stop it!

JR

**Comment by lelekos** [comment]
Hello,

I have actually made it to load the strings though a .txt file:

SecRule ARGS "@pmFromFile c:\inetpub\wwwroot\private\badstrings.txt" "t:normalisePathWin,deny,id:49459,severity:4,msg:'NO'"

@void-in thanks.

Alex

**Comment by md2k** [comment]
Hi guys, i have problem with ModSecurity, it is fail to parse `Content-Type: application/x-www-form-urlencoded` with error: `Unknown request body processor: application/x-www-form-urlencoded; charset=UTF-8`

Example of request is (it admin-ajax.php request in WP, one of them): 
```
curl 'https://www.domain.net/wp-admin/admin-ajax.php' -H 'Pragma: no-cache' -H 'Origin: https://www.domain.net' -H 'Accept-Encoding: gzip, deflate, br' -H 'Accept-Language: en-US,en;q=0.8,ru;q=0.6' -H 'Authorization: Basic bWQyazprajtiaix2Znk=' -H 'Content-Type: application/x-www-form-urlencoded; charset=UTF-8' -H 'Accept: application/json, text/javascript, */*; q=0.01' -H 'Cache-Control: no-cache' -H 'X-Requested-With: XMLHttpRequest' -H 'Connection: keep-alive' -H 'Referer: https://www.domain.net/wp-admin/admin.php?page=w3tc_dashboard' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.51 Safari/537.36' --data 'data%5Bwp-auth-check%5D=true&interval=60&_nonce=a4ca997c17&action=heartbeat&screen_id=toplevel_page_w3tc_dashboard&has_focus=false' --compressed
```

Audit log is:
```
{
    "transaction": {
        "client_ip": "111.111.111.111",
        "time_stamp": "Thu Jan 12 17:37:17 2017",
        "server_id": "c453bc20354668790965b09c5b033295f1e98e59",
        "client_port": 65043,
        "host_ip": "111.111.111.111",
        "host_port": 443,
        "id": "148423903790.931236",
        "request": {
            "protocol": "POST",
            "http_version": 1.1,
            "uri": "/wp-admin/admin-ajax.php",
            "headers": {
                "Host": "www.domain.net",
                "Connection": "keep-alive",
                "Content-Length": "129",
                "Pragma": "no-cache",
                "Cache-Control": "no-cache",
                "Authorization": "Basic bWQyazprajtiaix2Znk=",
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "Origin": "https://www.domain.net",
                "X-Requested-With": "XMLHttpRequest",
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.51 Safari/537.36",
                "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
                "Referer": "https://www.domain.net/wp-admin/admin.php?page=w3tc_dashboard",
                "Accept-Encoding": "gzip, deflate, br",
                "Accept-Language": "en-US,en;q=0.8,ru;q=0.6",
                "Cookie": "removed"
            }
        },
        "response": {
            "http_code": 400,
            "headers": {
                "Connection": "close",
                "Server": "nginx/1.10.2",
                "Date": "Thu, 12 Jan 2017 16:37:17 GMT",
                "Content-Length": "575",
                "Content-Type": "text/html"
            }
        },
        "producer": {
            "modsecurity": "ModSecurity v3.0.0-alpha (Linux)",
            "connector": "ModSecurity-nginx v0.1.1-beta",
            "secrules_engine": "Enabled",
            "components": [
                "OWASP_CRS/3.0.0\""
            ]
        },
        "messages": [
            {
                "message": "Failed to parse request body.",
                "details": {
                    "match": "Matched \"Operator `@eq' with parameter `0' against variable `REQBODY_ERROR' (Value: `1' ) \" at REQBODY_ERROR",
                    "ruleId": "200002",
                    "file": "/etc/nginx/modsecurity.conf",
                    "lineNumber": "53",
                    "data": "Unknown request body processor: application/x-www-form-urlencoded; charset=UTF-8",
                    "severity": "2",
                    "ver": "",
                    "rev": "",
                    "tags": [

                    ],
                    "maturity": "0",
                    "accuracy": "0"
                }
            }
        ]
    }
}
```

Debug Level 9 for such request here: https://gist.github.com/md2k/b0e5acc42c153a3c8a66eda6f40feb75

I use current lates master of v3/master branch and ModSec Nginx dynamic module repository to create plugin (also latest)
Nginx 1.10.2
Ubuntu 14.04.5 LTS

Any idea what wrong with it, from my point of view it is normal request which is pretty common in web?

**Comment by md2k** [comment]
Any updates for this ?

**Comment by md2k** [comment]
@zimmerle can you take a look for this issue? (i use V3 branch with Nginx through ModSecurity-Nginx module), OWASP 3.0.0 
Thank you.

**Comment by zimmerle** [comment]
Hi @md2k,

Originally this issue was about a problem on ModSecurity version 2.x. As I can see you are having trouble on v3.x. Do you mind to open an specific issue for it?

**Comment by md2k** [comment]
Hi @zimmerle , moved ticket to #1330

**Comment by zimmerle** [comment]
No longer a concern in libModSecurity. Marking it as won't fix for 2.x. Further information about libModSecurity available here:
https://github.com/SpiderLabs/ModSecurity/tree/v3/master
